internal-log-fetcher/src/discovery.rs

```diff
@@ -42,30 +43,84 @@ fn new_from_aws() -> Result<DiscoveryService> {
     let s3 = AmazonS3Builder::from_env()
         .with_bucket_name(bucket)
         .build()?;
-    let aws = AwsConfig {
-        s3: s3,
-        prefix: prefix,
-    };
+    let aws = AwsConfig { s3, prefix };
     Ok(DiscoveryService {
         aws: Some(aws),
         online_url: None,
     })
 }
 
-async fn fetch_online(url: &str) -> Result<HashSet<NodeIdentity>> {
-    // Use reqwest to make an async GET request.
-    let response = reqwest::get(url).await?;
+/// Constructs a `NodeIdentity` based on the provided parameters.
+///
+/// - `remote_addr`: The remote address as a string.
+/// - `control_port`: The node GraphQL control port number.
+/// - `submitter`: Submitter's public key.
+/// - `host_overrides`: An optional list of template strings for host addresses overriding.
+///
+/// Returns a `NodeIdentity` struct.
+fn new_node_identity(
+    remote_addr: &str,
+    control_port: u16,
+    submitter: String,
+    host_overrides: Option<&[String]>,
+) -> NodeIdentity {
+    // In certain environment configurations, we want to override the host components
+    // of online fetched remote addresses with provided host override templates.
+    // Please refer to the main.rs::test_fetch_online() test for an example.
+    if control_port >= 10000 {
+        // The control port is in the format of 10000 + index.
+        let index = (control_port / 10000) as usize - 1;
+        if let Some(overrides) = host_overrides {
+            // Check if the index is within the bounds of the provided host overrides.
+            if let Some(url_template) = overrides.get(index) {
+                // Replace the placeholder "{}" with the control port modulo 10000.
+                let template_value = control_port % 10000;
+                let ip = url_template.replace("{}", &template_value.to_string());
+                // Return the NodeIdentity with the overridden host and default port 80.
+                return NodeIdentity {
+                    ip,
+                    graphql_port: 80,
+                    submitter_pk: Some(submitter),
+                };
+            }
+        }
+    }
 
+    // Try to parse the remote address as a socket address (AWS case).
+    if let Ok(socket_addr) = remote_addr.parse::<SocketAddr>() {
+        NodeIdentity {
+            ip: socket_addr.ip().to_string(),
```

Comment:
Why is this case necessary, why not always use `remote_addr.to_string()`?

Reply:
Is it replacement for previously-present logic?

```rust
let colon_ix = meta
            .remote_addr
            .find(':')
            .unwrap_or_else(|| meta.remote_addr.len());
```

Reply:
Yes, so the original logic from my understanding was to simply extract the host/IP from remote address fetched from AWS:

```rust
for (_, meta) in aws_results {
        let colon_ix = meta
            .remote_addr
            .find(':')
            .unwrap_or_else(|| meta.remote_addr.len());
        results.insert(NodeIdentity {
            ip: meta.remote_addr[..colon_ix].to_string(),
            graphql_port: meta.graphql_control_port,
            submitter_pk: Some(meta.submitter),
        });
    }
```

I replaced it with `SocketAddr` parsing to preserve the logic since I thought that addresses from AWS might be returned in format of `<IP:PORT>` which is different to what Uptime backend returns.

In the end we default to `remote_addr.to_string()`.

Reply:
Okay, just to confirm: both `245.123.1.89` and `245.123.1.89:35031` will be accepted for `remote_addr` with the resulting `ip` ultimately being `245.123.1.89`?

Reply:
Yes, exactly.